epytope/IO/MartsAdapter.py

```diff
@@ -7,107 +7,338 @@
+        Create the biomart header with given properties
+
+        :param dict mart_properties: The fields and values for the header 
+
+        :return: The biomart header as XML
+        :rtype: ElementTree.Element
+        """
+        # XML header will be added before query is sent
+        biomart_header = ElementTree.Element("Query")
+        biomart_header.attrib.update(mart_properties)
+        return biomart_header
+
+    def __add_attribute(self, parent, attribute_name, attribute_value, name="Attribute"):
+        """
+        Adds an attribute to the ElementTree SubElement
+
+        :param ElementTree parent: The parent element (which is getting modified)
+        :param str attribute_name: The name of the attribute
+        :param str attribute_value: The value of the attribute
+        :param str name: The name of the XML field (default: Attribute)
+
+        :return: The SubElement with attribute
+        :rtype: ElementTree.SubElement
+        """
+        element = ElementTree.SubElement(parent, name)
+        element.set(attribute_name, attribute_value)
+        return element
+
+    def __add_filter(self, parent, filter_id, id, filter_value, value, name="Filter"):
+        """
+        Adds a filter to the ElementTree SubElement
+
+        :param ElementTree parent: The parent element (which is getting modified)
+        :param str filter_id: The name of the filter name field
+        :param str id: The id of the filter
+        :param str filter_value: The name of the filter value field
+        :param str value: The value of the filter
+        :param str name: The name of the XML field (default: Attribute)
+
+        :return: The SubElement with attribute
+        :rtype: ElementTree.SubElement
+        """
+        element = ElementTree.SubElement(parent, name)
+        element.set(filter_id, id)
+        element.set(filter_value, value)
+        return element
+
+    def __search_for_resources(self, xml_root):
+        """
+        Perform search (GET) using the given XML as query value
+
+        :param ElementTree Element xml_root: The XML of the query
+
+        :return: The found resources as data frame
+        :rtype: pandas.core.frame.DataFrame
+        """
+        f = io.BytesIO()
+        et = ElementTree.ElementTree(xml_root)
+        et.write(f, encoding="UTF-8", xml_declaration=True)
+        response = self.__perform_request(self.biomart_url, {"query": f.getvalue().decode("utf8")})
+        result = pd.read_csv(io.StringIO(response.content.decode('utf-8')), delimiter='\t')
+        return result
+
+    def __chunks(self, lst, n):
+        for i in range(0, len(lst), n):
+            yield lst[i:i + n]
+
+    def __perform_request(self, url, params):
+        """
+        Perform a request (GET) on the given url and parameters. 
+        Request will be retried three times on specified errors.
+
+        :return: The response
+        :rtype: request.response
+        """
+        retry_strategy = Retry(
```

Comment:
I also checked, it seems like you mentioned that it's likely due to the requests rate as EBI services have a [rate limit](https://www.ebi.ac.uk/training/online/sites/ebi.ac.uk.training.online/files/u1218/8_REST_API_Emily.pdf). It seems like if you get a `429` response then you should wait 40s before retry. So should there be set a 40s waiting time before next retry?

Reply:
So to try after 40s the backof should be at least 6 for what I've calculated here: https://urllib3.readthedocs.io/en/stable/reference/urllib3.util.html